### PR TITLE
Use Huffman compression for blackbox data transfer

### PR DESCRIPTION
**Pull request by martinbudden**
Will require configurator support.

### CODE REVIEW

File: src/main/common/huffman.c [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+
+int huffmanEncodeBuf(uint8_t *outBuf, int outBufLen, const uint8_t *inBuf, int inLen, const huffmanTable_t *huffmanTable)
+{
+    int ret = 0;
+
+    uint8_t *outByte = outBuf;
+    *outByte = 0;
+    uint8_t outBit = 0x80;
+    for (int ii = 0; ii< inLen; ++ii) {
```
Comment (DieHertz) [inline src/main/common/huffman.c]: Seems like `ii` and `inBuf` are redundant, I would rewrite this loop as:
```
for (const uint8_t *inBufEnd = inBuf + inLen; inBuf < inBufEnd; ++inBuf)
```
or at least move the inBuf increment into for loop as it semantically belongs there:
```
for (int ii = 0; ii < inLen; ++ii, ++inBuf)
```

Comment (martinbudden) [reply src/main/common/huffman.c]: Yes, your first suggestion is better than my code.

File: src/test/unit/huffman_unittest.cc [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+            huffManLenIndex[huffmanTree[ii].codeLen] = ii;
+        }
+    }
+}
+
+int huffmanDecodeBuf(uint8_t *outBuf, int outBufLen, const uint8_t *inBuf, int inBufLen, int inBufCharacterCount, const huffmanTree_t *huffmanTree)
+{
+    static bool initialized = false;
+    if (!initialized) {
+        huffmanInitDecodeLenIndex();
```
Comment (DieHertz) [inline src/test/unit/huffman_unittest.cc:326]: It's a bit inconsistent that this function can be supplied with a custom `huffmanTree` but `huffmanInitDecodeLenIndex` assumes the default one.

**Comment by digitalentity** [comment]
Will require corresponding change to Configurator plus increased MSP API version since we change the MSP command format

**Comment by martinbudden** [comment]
Indeed. This is now supported in the Betaflight configurator, so porting the code across should be straightforward.

**Pull request closed without merging**